### DISCUSSION TITLE
Roadmap to 1.0.0

### DISCUSSION THREAD

**Discussion by eloquence** [freedomofpress/dangerzone#539] [Ideas]
We're not close to a 1.0.0 release yet, but I would suggest we look ahead as to what we would expect from such a release. Typically, a 0.x release series is used to denote a product still in active development, with breaking changes potentially introduced at any time, and core functionality still missing.

To me, a 1.0.0 milestone would imply a few things:
- platform support maturity across Windows/Mac/Linux/Qubes
- improved installation (ideally, smaller initial download size; if possible, avoiding the separate Docker installation requirement)
- a polished GUI that mitigates user error through clear [affordances](https://en.wikipedia.org/wiki/Affordance), by applying standard UX patterns like [progressive disclosure](https://en.wikipedia.org/wiki/Progressive_disclosure), etc.
- a public threat model (potentially augmented by a more well-defined internal one)
- a public architecture diagram
- ideally, a completed independent security audit, highest priority findings addressed
- a more polished visual identity

What's missing from that list? What's in it that maybe doesn't need to be in a 1.0.0 release?

**Comment by apyrgio** [comment]
I've been revisiting this post on and off,  because while I like each item in the list, I find myself not being 100% sure if that constitutes Dangerzone 1.0.0. I mean, to my mind, achieving the above is a definite sign that Dangerzone is **stable** and **mature**. A 1.0.0 release is of course tangent to these concepts, but to me, it's mainly about setting expectations.

It means "From now on, I will make it clear when I break your workflow, or when I introduce a new feature". This applies both to the UX and the API. And implies that "I do not plan to make radical changes right away". That's at least how I interpret [SemVer's](https://semver.org/) wording on this subject:

> **How do I know when to release 1.0.0?**
> 
> If your software is being used in production, it should probably already be 1.0.0. If you have a stable API on which users have come to depend, you should be 1.0.0. If you’re worrying a lot about backward compatibility, you should probably already be 1.0.0.

_(see also this interesting discussion: https://github.com/semver/semver/issues/734)_

---

With this angle in mind, I'll comment on the above points:

> platform support maturity across Windows/Mac/Linux/Qubes :+1: 

I guess this mainly applies to Qubes, and how users install Dangerzone and update it. So yes, I agree that having Qubes support on par with other systems is a 1.0.0 requirement.

> improved installation (ideally, smaller initial download size; if possible, avoiding the separate Docker installation requirement) :thinking: 

I think that shrinking the image is not a 1.0.0 requirement (even though I _hate_ our large images). Using a different container runtime than Docker on MacOS/Windows (if we do choose to go down that road) is indeed a major version bump (either 0 -> 1 or 1 -> 2). But if this is not something that is in our immediate plans, I would not make this a strict requirement.

> a polished GUI that mitigates user error through clear [affordances](https://en.wikipedia.org/wiki/Affordance), by applying standard UX patterns like [progressive disclosure](https://en.wikipedia.org/wiki/Progressive_disclosure), etc. :thinking: 

This is a pretty interesting item, and a thorn against my reading of SemVer 1.0.0. Our GUI does not constitute an API, yet it's the main way people use Dangerzone. SemVer does not take this into account. Still, since we've broaden the concept of SemVer to "breaking expectations" than just "breaking programmatic interface", I'd suggest that we treat any breaking change to the way user's use our UI as a major version bump. Some examples are changing the existing workflow to a large extent, or making a major UX overhaul.

So, is this item a 1.0.0 requirement? I'd say only if we plan to change our sanitization workflow (may happen if we want to target document previewing now, else no) or if we radically change our UX (this is part of our short term plans, so maybe yes).

> a public threat model (potentially augmented by a more well-defined internal one) 
> a public architecture diagram

I see these as a good to have, but I don't think  they are a 1.0.0 requirement.

> ideally, a completed independent security audit, highest priority findings addressed :+1: 

if the security audit finds something that undermines the way we perform sanitization and requires radical changes, I agree. This is also part of our short-term plans, so we'll find out soon.

>  a more polished visual identity

Again, this seems more of a good to have for me.

---

Finally, some honorary mentions:

1. Stabilizing the `dangerzone-cli` interface. This interface has not changed for a while, and there are people who already depend on it. I think we are safe to call it stable.
2. Creating a stable Python API for integrating Dangerzone as a sanitization *library*. People currently use Dangerzone as a separate process (`dangerzone-cli`), but not as a Python module (`from dangerzone import ...`). Having a Python API for Dangerzone, and uploading it to PyPI, would make things more interesting for adoption, and it's where SemVer would fit as well.

**Comment by deeplow** [comment]
Maybe also localization should be a requirement https://github.com/freedomofpress/dangerzone/issues/665

**Comment by eloquence** [comment]
Some updates:
- First audit was completed last year: https://freedom.press/tech/news/dangerzone-receives-favorable-audit/
- The audit and general maintenance experience has led to us prioritizing independent container updates as a must-have requirement on the road to 1.0.0; this is currently in the works (https://github.com/freedomofpress/dangerzone/pull/1080)
- Thanks to an engagement with Superbloom and Okthanks, we have a set of UX recommendations that we can further refine and implement: https://github.com/freedomofpress/dangerzone/issues?q=state%3Aopen%20label%3A%22gui-v2%22 - probably only a couple of select changes would make it into 1.0.0.
- Prep for threat model work is underway.

**Comment by harrislapiroff** [comment]
We had a good conversation this morning about what constitutes a 1.0 release. We can't really go by semver here since semver isn't well suited to GUI applications (what constitutes a breaking change to a GUI). Arguably, more than any technical milestone, 1.0 represents more of a marketing milestone for DZ—an opportunity to encourage new users to install the application and to minimize the likelihood they bounce off of it.

We came up with a draft list that bears some similarities to @eloquence's list above, though we tried to take a minimal approach to what *must* be included. We'll continue to refine and add to this list later this week:

* Confidence in its security (✅ recent security audit) 
* Defined and documented platform support (✅ [install.md](https://github.com/freedomofpress/dangerzone/blob/main/INSTALL.md#operating-system-support))
* Defined container runtime support (do we give users options for container runtimes and which runtimes under which circumstances?)
* Nice to have:
  * Polished conversion flow UX
  * Stabilized and documented DZ CLI API
  * Embedded Podman
* Independent Container Updates (in progress)
* Public Roadmap
* "Complete" public documentation

**Comment by harrislapiroff** [comment]
Here's a draft roadmap:

## 0.10.0

**Primary Features:** Independent container updates, embedded container runtime (Podman)

- [ ] Public Roadmap (that's this!)
- [ ] #1006 
- [ ] #1151
- [ ] #1145

## ~0.11.0~

> [!note]
> As of **2025-05-20** we determined that it's not really practical to maintain Docker support alongside independent container updates, which means that instead of having a 0.11.0 release where embedded Podman is the primary focus, we need to roll that functionality into 0.10.0

~Do we need this as a separate release? Should we consolidate it with 1.0? Because ICU will have landed, there's less pressure to issue regular releases for security reasons.~

~On the other hand, issuing regular releases still has a security benefit for people who decline container updates (e.g., because they are airgapped) as well as for updating other dependencies like `cosign` and `podman`.~

## 1.0

**Primary features:** Polished GUI, improved documentation

- [x] Recent security audit
- [x] Platform support documentation
- [ ] #903
- [ ] #213 
- [ ] #132 
- [ ] #894 

## Planned

- [ ] DZ distributions
  - [ ] #1081 
  - [ ] #1007 
- [ ] Localization to specific languages
- [ ] #633  
- [ ] Website Improvements

## Possible Future

- [ ] Python API
- [ ] Document container image (for people to use the image without using the DZ app)
- [ ] DZ server
  - [ ] DZ mobile clients
- [ ] DZ preview/viewer 
- [ ] #1142 
- [ ] DZ forum/chat?
- [ ] File Formats
  - [ ] #1030 
  - [ ] #760 
  - [ ] #1148 
- [ ] OCR improvements
  - [ ] #407 
  - [ ] #403 
- [ ] #1069 
- [ ] Provide document metadata
- [ ] Explore options for making DZ a default app for certain file types
- [ ] Watch folder option
- [ ] #165
